### PR TITLE
feat(desktop): iMessage-style time grouping in chat (NAN-680)

### PR DESCRIPTION
**Pull request by yagudaev**
## Why

Long voice conversations produce a noisy chat transcript when every bubble carries its own timestamp. iMessage solves this by grouping nearby messages into bursts and only labeling boundaries — same speaker, same minute, one separator. This PR brings that pattern to VoiceClaw desktop chat, surfaces clearer day boundaries, and exposes timing metadata to the brain agent so the model can answer time-relative questions ("what did we talk about 5 minutes ago?") from structured data instead of guessing from ordering.

Reference iMessage screenshot the user shared: https://files.openclaw.ai/inbound/file_624---52fce8b0-3a9f-45b9-88ca-841699b57155.jpg

## What

- New pure helper `desktop/src/renderer/src/lib/message-grouping.ts` exporting `groupMessages(messages, options)` that returns separators + per-message burst metadata. Burst threshold is 60s (configurable via `burstThresholdMs`); same role + same day required to stay in a burst.
- Day separators: `Today 9:14 AM` / `Yesterday 10:42 PM` / `Tuesday 8:03 AM` / `Mar 12 4:51 PM`. Intra-day burst separators: `just now` / `5 min ago` / `2 hr ago` / clock time for older.
- New `MessageGroupSeparator` component renders separators with subtle styling (uppercase, low contrast, centered) — palette matches existing muted-foreground.
- `MessageBubble` accepts `showTimestamp` (per-bubble inline stamp) and `isLastInBurst` (tightens spacing inside a burst), plus a `title` tooltip with the exact timestamp on hover.
- Reused the NAN-679 `MessageContextMenu` to add a `Show times` / `Hide times` item that toggles a conversation-wide `showTimes` state in `ChatPage`. Conversation-wide rather than per-message because when you want to reconcile timestamps you almost always want them on every message in view.
- `conversationHistory` payload sent to the relay now carries optional `timestamp` and `relativeMs` per turn. Relay's `formatRecentTurnsPreamble` prefixes each verbatim line with `[3m ago]` / `[2h ago]` / `[1d ago]` when relative time is present and adds a one-line instruction telling the model to use those stamps to answer time-relative questions. Backward compatible — older clients that don't send the fields keep the unstamped preamble.
- Vitest cases for `groupMessages` cover: empty, single, two-30s-apart-same-burst, two-90s-apart-different-bursts, day boundary, role transition, threshold override.
- Starlight docs page `docs/src/content/docs/desktop/time-grouping.mdx` (user-facing copy — explains grouping, separators, hover tooltip, Show times toggle) wired into the Guides sidebar.

## Screenshots

Reference: iMessage grouping

## Test plan

- [x] `yarn workspace voiceclaw-desktop typecheck` — green
- [x] `yarn workspace voiceclaw-desktop test` — 66 of 66 tests pass; the only failing suite (`diagnostic-bundle.test.ts`) was already broken on `origin/main` because `archiver` is required at runtime but not declared in `desktop/package.json`. Unrelated to this PR.
- [x] `yarn workspace voiceclaw-desktop build` — green
- [x] `yarn workspace voiceclaw-relay-server test` — 72 of 72 pass (2 skipped, pre-existing)
- [x] `yarn workspace voiceclaw-relay-server build` — green
- [x] `yarn workspace voiceclaw-docs build` — green; new `time-grouping` page included in sidebar
- [ ] Manual on macOS: open an existing conversation; confirm bursts collapse together inside the 60s window and split when the speaker switches; check `Today` / `Yesterday` / weekday / date separators against a multi-day conversation
- [ ] Manual: right-click a bubble, choose `Show times`; confirm every message gets a small timestamp; pick `Hide times` to collapse
- [ ] Manual: hover any bubble; tooltip shows the exact send time
- [ ] Manual: start a call after a few minutes idle; ask "what did we talk about a few minutes ago?" and confirm the brain references the prior turn from the `[Xm ago]` stamps

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: desktop/src/renderer/src/lib/message-grouping.ts [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+export type GroupedItem =
+  | { kind: 'separator', label: string, timestamp: number }
+  | {
+      kind: 'message'
+      message: Message
+      isFirstInBurst: boolean
+      isLastInBurst: boolean
+      showTimestamp: boolean
```
Comment (yagudaev) [inline desktop/src/renderer/src/lib/message-grouping.ts]: Dead field — `showTimestamp` is always set to `false` (line 74) and is never read from `GroupedItem` anywhere in ChatPage. The `showTimes` React state is passed directly to `MessageBubble` as a prop instead. This field pollutes the public type for no reason; remove it from `GroupedItem` and from line 74.

File: desktop/src/renderer/src/pages/ChatPage.tsx [hunk 2]
```diff
@@ -315,11 +322,17 @@ export function ChatPage() {
...
     const conversationHistory = convId
       ? (await getMessages(convId))
           .filter((m) => m.role === 'user' || m.role === 'assistant')
           .slice(-200)
-          .map((m) => ({ role: m.role as 'user' | 'assistant', text: m.content }))
+          .map((m) => ({
+            role: m.role as 'user' | 'assistant',
+            text: m.content,
+            timestamp: m.created_at,
+            relativeMs: Math.max(0, nowMs - m.created_at),
```
Comment (yagudaev) [inline desktop/src/renderer/src/pages/ChatPage.tsx]: `relativeMs` is snapped to `nowMs = Date.now()` at the moment `startCall` fires. If there's any async latency (device enumeration, DB read, relay handshake) between `nowMs` capture and the relay actually injecting the preamble, all stamps are off by that delta. Low-severity for typical latency (<1s), but a stale `nowMs` snapshot at the top of the async chain means the relay sees slightly wrong ages. Consider passing `timestamp` only and computing `relativeMs` server-side at inject time, or at least move `nowMs` assignment to just before `realtime.start()` is called rather than 10+ async lines earlier.

File: relay-server/src/history.ts [hunk 3]
```diff
@@ -54,13 +59,15 @@ export function formatRecentTurnsPreamble(recent: HistoryMessage[]): string {
       const text = typeof m.text === "string" ? m.text.trim() : ""
       if (!text) return null
       const speaker = m.role === "assistant" ? "Assistant" : "User"
-      return `${speaker}: ${text}`
+      const stamp = formatRelativeStamp(m.relativeMs)
+      return stamp ? `${stamp} ${speaker}: ${text}` : `${speaker}: ${text}`
```
Comment (yagudaev) [inline relay-server/src/history.ts:68]: No relay-side unit test covers `formatRelativeStamp` or the updated `formatRecentTurnsPreamble` (stamp present, stamp absent, boundary at 0s/60s/3600s). The existing `gemini-history-inject` suite passes because it uses history entries without `relativeMs`, so the new code path is entirely untested. A test like: `formatRecentTurnsPreamble([{ role: 'user', text: 'hi', relativeMs: 300_000 }])` should contain `[5m ago] User: hi`.

File: docs/src/content/docs/desktop/time-grouping.mdx [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+Show times is a single conversation-wide toggle, not per-message — when you want timestamps, you almost always want them on every message in view.
+
+<Aside type="note">
+The toggle resets when you start a new conversation or restart the app. It's a transient view setting, not a saved preference.
+</Aside>
+
+## What the agent sees
+
+The grouping is purely a UI concern — it doesn't change what the agent receives. But the timing metadata is exposed to the agent so it can answer questions like _"what did we talk about five minutes ago?"_ Each historical turn the agent sees is tagged with a relative timestamp (`[3m ago]`, `[2h ago]`, `[1d ago]`) inside the conversation preamble. That's enough for the model to answer time-relative questions without having to guess from the transcript order alone.
```
Comment (yagudaev) [inline docs/src/content/docs/desktop/time-grouping.mdx]: Audience-separation concern (feedback_docs_audience.md — hard rule). The last paragraph exposes the `[3m ago]` wire-format stamp and "conversation preamble" — implementer-level details. A user doesn't need to know the internal notation the relay injects; they only need to know that the agent can answer time-relative questions. Suggested rewrite: drop the backtick examples and the word "preamble" — keep only the user-visible consequence ("the agent can tell you what happened five minutes ago").

File: desktop/src/renderer/src/lib/message-grouping.ts [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+function formatBurstSeparator(ts: number, now: number): string {
+  const diffMs = now - ts
+  if (diffMs < 60_000) return 'just now'
+  const diffMin = Math.round(diffMs / 60_000)
+  if (diffMin < 60) return `${diffMin} min ago`
+  const diffHr = Math.round(diffMs / 3_600_000)
+  if (diffHr < 6) return `${diffHr} hr ago`
+  return new Date(ts).toLocaleTimeString(undefined, { hour: 'numeric', minute: '2-digit' })
```
Comment (yagudaev) [inline desktop/src/renderer/src/lib/message-grouping.ts:150]: Edge case: a `diffMs` of 29_999ms (just under 30s) rounds to `1 min ago` via `Math.round(29999/60000) = 0` — wait, actually 29999/60000 = 0.5, rounds to 0 → `0 min ago`. That's a visible glitch: a message 30s old shows as '0 min ago' not 'just now'. The 'just now' guard is `diffMs < 60_000`, so messages 30–59s old are caught, but messages at exactly 30s get `Math.round(30000/60000) = Math.round(0.5) = 1` in most JS engines → '1 min ago' for a 30-second-old message. Consider using `Math.floor` instead of `Math.round` for the minute / hour buckets to avoid premature escalation.

**Review by yagudaev — COMMENTED** [review]
Solid feature overall — the grouping logic is correct, timezone handling uses `getFullYear/Month/Date` (local-time, no UTC drift), and the burst/day-boundary separation works properly. The pure-helper pattern for `message-grouping.ts` is clean and the memoized `timelineItems` avoids unnecessary re-renders. The two hard-rule violations need fixing before merge.

## BLOCKERs

**1. Docs audience-separation violation** (`docs/src/content/docs/desktop/time-grouping.mdx` line 58)
The "What the agent sees" paragraph exposes `[3m ago]` / `[2h ago]` / `[1d ago]` wire-format literals and the word "conversation preamble". These are implementer details that belong in the PR body, not the user-facing docs site. Drop the backtick examples and preamble terminology; keep only the user-visible outcome: "you can ask the agent what happened a few minutes ago and it will know."

**2. Dead `showTimestamp` field on `GroupedItem`** (`message-grouping.ts` line 15, 74)
The field is always `false` and is never consumed — `ChatPage` passes `showTimes` state directly to `MessageBubble`. Remove the field from the type and the push on line 74.

## Non-blocking issues

**`Math.round` premature escalation in `formatBurstSeparator`** (`message-grouping.ts` line 152)
A message exactly 30s old hits `Math.round(0.5) = 1` → "1 min ago" despite being just under a minute. `Math.floor` is the right choice for "minutes elapsed."

**`relativeMs` snapshot drift** (`ChatPage.tsx` line 325)
`nowMs` is captured early in the async `startCall` chain, several awaits before `realtime.start()`. Device enumeration + DB read can take 100–500ms; all stamps end up slightly stale. Low-severity but easy to fix: move `nowMs` to just before the `.map()` call, or better yet compute `relativeMs` from `timestamp` server-side at preamble-inject time.

## Test coverage gaps

- No relay-side test for `formatRelativeStamp` or the updated `formatRecentTurnsPreamble` with stamps. The new `[Xm ago]` code path is entirely uncovered — add a `history.test.ts` with stamp-present / stamp-absent / boundary cases (0s, 59s, 60s, 3599s, 3600s).
- `formatExactTimestamp` tests only cover same-day and yesterday. Missing: >7 days (full date), same year vs. different year, and a ts *in the future* (defensive).
- No test for the `diffDays <= 0` branch in `formatDaySeparator` when `ts` is slightly in the future (e.g., a message created 1ms from now due to clock skew).

## Code style

No ticket IDs or commit-message paragraphs in source comments on any PR-touched file. Clean.

No semicolons, helpers at bottom — compliant.

## Inline comments posted: 5

**Comment by yagudaev** [comment]
Addressed both BLOCKERs and both nits in the latest commit.

1. **`time-grouping.mdx:58`** — rewrote the "What the agent sees" paragraph to drop `[3m ago]`/`[2h ago]`/`[1d ago]` literals and "conversation preamble." Now reads as user-facing context only.
2. **Dead `showTimestamp` on `GroupedItem`** — removed from the type and the producer; `ChatPage` was always passing the React `showTimes` state directly to `MessageBubble`, so nothing else needed to change. `MessageBubble`'s `showTimestamp` prop stays (it's the actual display control).
3. **Math.round → Math.floor** in `formatBurstSeparator` so a 90s-old message no longer reports "2 min ago." Same fix applied in `relay-server/src/history.ts:formatRelativeStamp`.
4. **`nowMs` drift** — fixed by moving the relative-time calculation server-side. Desktop now sends `timestamp` only; relay derives `relativeMs` at format time via `Date.now()`. No async-gap drift possible.

Verified: typecheck green; desktop 66/66 + relay 72/72 tests pass; the pre-existing `diagnostic-bundle.test.ts` failure on origin/main is unchanged.
